blockchain/blockchain.go

```diff
@@ -1434,3 +1443,33 @@ func (b *Blockchain) GetBlockByNumber(blockNumber uint64, full bool) (*types.Blo
 func (b *Blockchain) Close() error {
 	return b.db.Close()
 }
+
+// CalculateBaseFee calculates the basefee of the header.
+func (b *Blockchain) CalculateBaseFee(parent *types.Header) uint64 {
+	if !b.config.Params.Forks.IsLondon(parent.Number) {
+		return chain.GenesisBaseFee
+	}
+
+	parentGasTarget := parent.GasLimit / elasticityMultiplier
+
+	// If the parent gasUsed is the same as the target, the baseFee remains unchanged.
+	if parent.GasUsed == parentGasTarget {
+		return parent.BaseFee
+	}
+
+	// If the parent block used more gas than its target, the baseFee should increase.
+	if parent.GasUsed > parentGasTarget {
+		gasUsedDelta := parent.GasUsed - parentGasTarget
+		y := parent.BaseFee * gasUsedDelta / parentGasTarget
+		baseFeeDelta := y / defaultBaseFeeChangeDenom
+
+		return parent.BaseFee + common.Max(baseFeeDelta, 1)
+	}
+
+	// Otherwise, if the parent block used less gas than its target, the baseFee should decrease.
+	gasUsedDelta := parentGasTarget - parent.GasUsed
+	y := parent.BaseFee * gasUsedDelta / parentGasTarget
+	baseFeeDelta := y / defaultBaseFeeChangeDenom
```

Comment:
This code is duplicated, would be great to maintain these formulas in a function

Reply:
Done